File: frontend/types/analytics.ts
/**
 * Phase 9 — Analytics, Predictions, and Reports TypeScript types.
 *
 * Mirrors the backend Pydantic schemas exactly so the compiler catches
 * any API contract drift at build time.
 */

// ---------------------------------------------------------------------------
// Shared primitives
// ---------------------------------------------------------------------------

export interface MonthlyCount {
  month: string;
  count: number;
}

export interface MonthlyRevenue {
  month: string;
  revenue: number;
  expenses: number;
  profit: number;
}

export interface RoleCount {
  role: string;
  count: number;
}

export interface StatusCount {
  status: string;
  count: number;
}

// ---------------------------------------------------------------------------
// KPI Summary
// ---------------------------------------------------------------------------

export interface KPISummary {
  total_employees: number;
  total_documents: number;
  active_workflows: number;
  total_ai_conversations: number;
  completed_workflow_executions: number;
  total_predictions_run: number;
  total_reports_generated: number;
}

// ---------------------------------------------------------------------------
// Revenue Analytics
// ---------------------------------------------------------------------------

export interface RevenueAnalytics {
  year: number;
  data: MonthlyRevenue[];
  total_expenses_ytd: number;
  avg_monthly_expenses: number;
}

// ---------------------------------------------------------------------------
// Department Analytics
// ---------------------------------------------------------------------------

export interface DepartmentAnalyticsItem {
  department_id: string;
  department_name: string;
  employee_count: number;
  avg_salary: number;
  document_count: number;
  active_workflows: number;
}

export interface DepartmentAnalytics {
  data: DepartmentAnalyticsItem[];
  total_departments: number;
}

// ---------------------------------------------------------------------------
// Employee Analytics
// ---------------------------------------------------------------------------

export interface EmployeeAnalytics {
  total: number;
  active: number;
  new_this_month: number;
  role_distribution: RoleCount[];
  monthly_headcount: MonthlyCount[];
}

// ---------------------------------------------------------------------------
// Document Analytics
// ---------------------------------------------------------------------------

export interface DocumentAnalytics {
  total_uploads: number;
  monthly_uploads: MonthlyCount[];
  by_status: StatusCount[];
  by_ocr_status: StatusCount[];
}

// ---------------------------------------------------------------------------
// Workflow Analytics
// ---------------------------------------------------------------------------

export interface WorkflowAnalytics {
  total_executions: number;
  by_status: StatusCount[];
  avg_duration_seconds: number;
  step_action_frequency: StatusCount[];
}

// ---------------------------------------------------------------------------
// AI Usage Analytics
// ---------------------------------------------------------------------------

export interface AIUsageAnalytics {
  total_conversations: number;
  monthly_conversations: MonthlyCount[];
  avg_response_time_seconds: number;
  total_prompt_tokens: number;
  total_completion_tokens: number;
}

// ---------------------------------------------------------------------------
// Predictions
// ---------------------------------------------------------------------------

export type PredictionModelName =
  | "revenue_forecast"
  | "employee_attrition"
  | "customer_churn";

export interface PredictionRequest {
  model: PredictionModelName;
  input?: Record<string, unknown>;
}

export interface Prediction {
  id: string;
  model: string;
  input?: Record<string, unknown> | null;
  prediction?: Record<string, unknown> | null;
  confidence?: number | null;
  execution_time?: number | null;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Reports
// ---------------------------------------------------------------------------

export type ReportTypeName =
  | "revenue"
  | "employees"
  | "departments"
  | "workflows"
  | "documents"
  | "ai_usage";

export type ReportFormatName = "csv" | "xlsx" | "pdf";

export type ReportStatus = "pending" | "completed" | "failed";

export interface ReportGenerateRequest {
  report_type: ReportTypeName;
  format?: ReportFormatName;
  filters?: Record<string, unknown>;
}

export interface Report {
  id: string;
  report_type: string;
  format?: string | null;
  status: ReportStatus;
  file_path?: string | null;
  generated_by?: string | null;
  generated_at?: string | null;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/analytics.py
"""
Analytics Pydantic schemas — Phase 9.

All response models for the /api/v1/analytics/* endpoints.

Design decisions:
  - All monetary values are float (Python) rather than Decimal to keep
    JSON serialisation simple; analytics are approximations, not accounting.
  - month labels use 3-letter abbreviations ("Jan", "Feb", …) for chart
    axis labels.
  - Optional fields use sensible defaults so empty-database queries never
    return None where the frontend expects a number.
"""

from __future__ import annotations

import uuid

from pydantic import BaseModel, Field

# ---------------------------------------------------------------------------
# Shared primitives
# ---------------------------------------------------------------------------


class MonthlyCount(BaseModel):
    """A (month-label, count) data point for trend charts."""

    month: str = Field(..., description="3-letter month abbreviation: 'Jan'…'Dec'.")
    count: int = Field(0, ge=0)


class MonthlyRevenue(BaseModel):
    """Monthly revenue/expense data point for the revenue chart."""

    month: str = Field(..., description="3-letter month abbreviation.")
    revenue: float = Field(0.0, ge=0)
    expenses: float = Field(0.0, ge=0)
    profit: float = Field(0.0)


class RoleCount(BaseModel):
    """Role name + member count for the role-distribution chart."""

    role: str
    count: int = Field(0, ge=0)


class StatusCount(BaseModel):
    """Generic status + count pair."""

    status: str
    count: int = Field(0, ge=0)


# ---------------------------------------------------------------------------
# KPI Summary (dashboard header)
# ---------------------------------------------------------------------------


class KPISummaryResponse(BaseModel):
    """Aggregated top-level KPIs for the analytics overview cards."""

    total_employees: int = 0
    total_documents: int = 0
    active_workflows: int = 0
    total_ai_conversations: int = 0
    completed_workflow_executions: int = 0
    total_predictions_run: int = 0
    total_reports_generated: int = 0


# ---------------------------------------------------------------------------
# Revenue Analytics
# ---------------------------------------------------------------------------


class RevenueAnalyticsResponse(BaseModel):
    """Monthly revenue trend for the selected year."""

    year: int
    data: list[MonthlyRevenue]
    total_expenses_ytd: float = 0.0
    avg_monthly_expenses: float = 0.0


# ---------------------------------------------------------------------------
# Department Analytics
# ---------------------------------------------------------------------------


class DepartmentAnalyticsItem(BaseModel):
    """Per-department breakdown row."""

    department_id: uuid.UUID
    department_name: str
    employee_count: int = 0
    avg_salary: float = 0.0
    document_count: int = 0
    active_workflows: int = 0


class DepartmentAnalyticsResponse(BaseModel):
    """Collection of department analytics items."""

    data: list[DepartmentAnalyticsItem]
    total_departments: int = 0


# ---------------------------------------------------------------------------
# Employee Analytics
# ---------------------------------------------------------------------------


class EmployeeAnalyticsResponse(BaseModel):
    """Employee trend and composition data."""

    total: int = 0
    active: int = 0
    new_this_month: int = 0
    role_distribution: list[RoleCount] = []
    monthly_headcount: list[MonthlyCount] = []


# ---------------------------------------------------------------------------
# Document Analytics
# ---------------------------------------------------------------------------


class DocumentAnalyticsResponse(BaseModel):
    """Document upload trends and status breakdown."""

    total_uploads: int = 0
    monthly_uploads: list[MonthlyCount] = []
    by_status: list[StatusCount] = []
    by_ocr_status: list[StatusCount] = []


# ---------------------------------------------------------------------------
# Workflow Analytics
# ---------------------------------------------------------------------------


class WorkflowAnalyticsResponse(BaseModel):
    """Workflow execution statistics."""

    total_executions: int = 0
    by_status: list[StatusCount] = []
    avg_duration_seconds: float = 0.0
    step_action_frequency: list[StatusCount] = []


# ---------------------------------------------------------------------------
# AI Usage Analytics
# ---------------------------------------------------------------------------


class AIUsageAnalyticsResponse(BaseModel):
    """AI conversation usage trends."""

    total_conversations: int = 0
    monthly_conversations: list[MonthlyCount] = []
    avg_response_time_seconds: float = 0.0
    total_prompt_tokens: int = 0
    total_completion_tokens: int = 0
